lib/actions/build.coffee

```diff
@@ -0,0 +1,224 @@
+			alias: 'c'
+		},
+		{
+			signature: 'application'
+			parameter: 'application'
+			description: 'The target resin.io application this build is for'
+			alias: 'a'
+		},
+		{
+			signature: 'tag'
+			parameter: 'tag'
+			description: 'The alias to the generated image'
+			alias: 't'
+		},
+		{
+			signature: 'nocache'
+			description: "Don't use cache when building"
+			boolean: true
+		},
+		{
+			signature: 'dockerHost'
+			parameter: 'dockerHost'
+			description: 'The address of the host containing the docker daemon'
+			alias: 'h'
+		},
+		{
+			signature: 'dockerPort'
+			parameter: 'dockerPort'
+			description: 'The port on which the host docker daemon is listening'
+			alias: 'p'
+		},
+		{
+			signature: 'ca'
+			parameter: 'ca'
+			description: 'TLS Certificate authority file'
+		},
+		{
+			signature: 'cert'
+			parameter: 'cert'
+			description: 'TLS Certificate file'
+		},
+		{
+			signature: 'key'
+			parameter: 'key'
+			description: 'TLS Key file'
+		}
+	]
+	action: (params, options, done) ->
+		# The default build context is the current directory
+		params.context ?= '.'
+
+		# Tar up the directory, ready for the build stream
+		tarDirectory(params.context)
+		.then (tarStream) ->
+			hooks =
+					buildSuccess: (image, layers) ->
+						if options.tag?
+							console.log("Tagging image as #{options.tag}")
+						done()
+					buildFailure: (err) ->
+						done("Build error: #{err.message}")
+					buildStream: (stream) ->
+						info = getBundleInfo(options)
+						# info is either a promise, or is undefined
+						if info?
+							info
+							.catch (err) -> done("Error: #{err.message}")
+							.then ([arch, deviceType]) ->
+								# Perform type resolution on the project
+								bundle = new resolve.Bundle(tarStream, deviceType, arch)
+								resolve.resolveBundle(bundle, resolve.getDefaultResolvers())
+							.then (resolved) ->
+								console.log("Building #{resolved.projectType} project")
+								# Send the resolved tar stream to the docker daemon
+								resolved.tarStream.pipe(stream)
+							.catch(done)
+						else
+							console.log('Warning: No architecture/device type or application information provided')
+							console.log('  No project type resolution will be performed.')
+							tarStream.pipe(stream)
+
+						# And print the output
+						stream.pipe(process.stdout)
+
+				# Create a builder
+				connectOpts = generateConnectOpts(options)
+
+				# Allow degugging output, hidden behind an env var
+				if process.env.RESIN_DEBUG_BUILD?
+					console.log('Connecting with the following options:')
+					console.log(JSON.stringify(connectOpts, null, '  '))
+
+				builder = new dockerBuild.Builder(connectOpts)
+				opts = {}
+
+				if options.tag?
+					opts['t'] = options.tag
+				if options.nocache?
+					opts['nocache'] = true
+				# Pass done as the error handler so any propagated errors
+				# hit the main handler
+				builder.createBuildStream(opts, hooks, done)
+		.catch(done)
```

Comment:
The flow to `done` here is quite confusing! The hooks call it, we also provide it as an error callback, and then we have a catch-all `catch` here that calls it even after that. I don't know what happens if we call done more than once, but it definitely feels like we shouldn't.

Can we do this with proper promises instead, and just cleanly map it back to the done callback at the end?

For example, if you have `new Promise((resolve, reject) =>` inside the `.then (tarStream)` here, you can call `resolve` or `reject` from the hooks, pass `reject` as the error handler for build.createBuildStream, and all this hooks business will turn back into just another step in this promise chain.

You can then put a single `nodeify(done)` at the end to go back to capitano. That handles success or any errors here or in the previous promise code, all in one place. If you start heading down that route, I expect there'll be more places where consistently using promises simplifies things a little too, although I can't see them off the top of my head.